main/src/library/Fixpoint3/Phase/IndexSelection.flix

```diff
@@ -0,0 +1,387 @@
+                    case _ =>
+                        MutList.push(exp, otherBoolExpList);
+                        acc
+                }, List.empty()) |> List.reverse |> List.toVector;
+            MutSet.add(rowVar, seenTuples);
+            let newBody = collectSearchesOp(body, indexes, seenTuples, rc);
+            MutSet.remove(rowVar, seenTuples);
+            collectSearch(relSym, search, indexes);
+            let otherBoolExps = MutList.toVector(otherBoolExpList);
+            let nestedBody =
+                if (Vector.length(otherBoolExps) == 0) newBody
+                else RelOp.If(otherBoolExps, newBody);
+            let equalExps = MutList.toVector(equalExpList);
+            if (Vector.length(equalExps) == 0)
+                RelOp.Search(rowVar, relSym, nestedBody)
+            else
+                RelOp.Query(rowVar, relSym, equalExps, -1, nestedBody)
+        case RelOp.Functional(rowVar, func, terms, body, arity) =>
+            MutSet.add(rowVar, seenTuples);
+            let res = RelOp.Functional(rowVar, func, terms, collectSearchesOp(body, indexes, seenTuples, rc), arity);
+            MutSet.remove(rowVar, seenTuples);
+            res
+        case RelOp.Project(_, _) => op
+        case RelOp.If(boolExps, body) => RelOp.If(boolExps, collectSearchesOp(body, indexes, seenTuples, rc))
+    }
+
+    ///
+    /// Collects the search `Search` for `relSym` to `indexes`.
+    ///
+    def collectSearch(relSym: RelSym, search: Search, indexes: Searches[r]): Unit \ r =
+        if (not Vector.isEmpty(search)){
+            let searches = MutMap.getWithDefault(relSym, List.empty(), indexes);
+            if (not List.memberOf(search, searches))
+                MutMap.put(relSym, search :: searches, indexes)
+            else ()
+        } else ()
+
+    ///
+    /// Returns true if `term` is ground (no free variables) w.r.t. `seenTuples`.
+    ///
+    def isTermGround(term: RamTerm, seenTuples: MutSet[RowVar, r]): Bool \ r = match term {
+        case RamTerm.Lit(_, _) => true
+        case RamTerm.RowLoad(rowVar, _, _) =>
+            MutSet.memberOf(rowVar, seenTuples)
+        case _ => false
+    }
+
+    mod AutomaticIndexSelection {
+        use Fixpoint3.Ast.Ram.Search
+        ///
+        /// This is based on the paper "Automatic Index Selection for Large-Scale Datalog
+        /// Computation" by Pavle Subotić, Herbert Jordan, Lijun Chang, Alan Fekete,
+        /// Bernhard Scholz (https://www.vldb.org/pvldb/vol12/p141-subotic.pdf)
+        ///
+        type alias Vertex = Int32
+        type alias Edge = (Vertex, Vertex)
+        // Matching is a set but we use a `Map` here so we can call `rangeQueryWith`
+        type alias Matching = Map[Edge, Unit]
+        type alias AdjacencyList = MultiMap[Vertex, Vertex]
+
+        ///
+        /// Find the minimum set of indexes that covers `searches`.
+        ///
+        @Internal
+        pub def minIndex(searches: List[Search]): List[Search] =
```

Comment:
Perhaps in "phase 2" we could consider whether a specific function like this one would benefit from unit testing. 

But let us leave it for now.